themes/powerline/powerline.base.sh

```diff
@@ -3,20 +3,60 @@
 # Define this here so it can be used by all of the Powerline themes
 THEME_CHECK_SUDO=${THEME_CHECK_SUDO:=false}
 
+function hex_to_rgb() {
+    local hex_color="$1"
+    local r g b
+
+    r=$(printf '%d' 0x${hex_color:1:2})
+    g=$(printf '%d' 0x${hex_color:3:2})
+    b=$(printf '%d' 0x${hex_color:5:2})
+
+    echo "${r};${g};${b}"
+}
+
 function set_color {
+  local fg bg fg_color bg_color
+
   if [[ "${1}" != "-" ]]; then
-    fg="38;5;${1}"
+    if [[ ${1} =~ ^[0-9]+$ ]]; then
+      fg_color="38;5;${1}"  # ANSI 256-color code
+    elif [[ ${1} =~ ^[0-9]{1,3}(\;[0-9]{1,3}){2}$ ]]; then
+      fg_color="38;2;${1}"  # RGB color code
+    elif [[ ${1} =~ ^#[0-9A-Fa-f]{6}$ ]]; then
+      fg_color="38;2;$(hex_to_rgb "${1}")"  # Hex color code converted to RGB
+    fi
   fi
+
+  # Check if background color is provided and is a number, RGB string, or hex code
   if [[ "${2}" != "-" ]]; then
-    bg="48;5;${2}"
-    [[ -n "${fg}" ]] && bg=";${bg}"
+    if [[ ${2} =~ ^[0-9]+$ ]]; then
+      bg_color="48;5;${2}"  # ANSI 256-color code
+    elif [[ ${2} =~ ^[0-9]{1,3}(\;[0-9]{1,3}){2}$ ]]; then
+      bg_color="48;2;${2}"  # RGB color code
+    elif [[ ${2} =~ ^#[0-9A-Fa-f]{6}$ ]]; then
+      bg_color="48;2;$(hex_to_rgb "${2}")"  # Hex color code converted to RGB
+    fi
+  fi
+
+  # Construct the color escape sequence
+  fg="${fg_color:-}"
+  bg="${bg_color:-}"
+  if [[ -n "$fg" && -n "$bg" ]]; then
+    # Both foreground and background are provided
+    echo -e "\[\033[${fg};${bg}m\]"
+  elif [[ -n "$fg" ]]; then
+    # Only foreground is provided
+    echo -e "\[\033[${fg}m\]"
+  elif [[ -n "$bg" ]]; then
+    # Only background is provided
+    echo -e "\[\033[${bg}m\]"
   fi
```

Comment:
Couldn't this be realized by just `${var:+...}`?

```suggestion
  echo -e "\[\033[${fg}${fg:+${bg:+;}}${bg}m\]"
```